### PR TITLE
qt 6.3.0

### PR DESCRIPTION
**Pull request by owine**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by danielnachun** [comment]
This currently fails because the patches no longer apply cleanly.  They could have been merged upstream and may not be needed.  Given how long this takes to run it would probably be good to figure out if those commits have been merged before trying anything else.

**Comment by owine** [comment]
I've removed the `assimp`-related patch as that was further updated upstream and no longer necessary.

**Comment by cho-m** [comment]
Cancelling run as patch does not apply:
```
==> Applying CVE-2022-1096-qtwebengine-6.3.diff
  (Stripping trailing CRs from patch.)
  patching file src/3rdparty/chromium/v8/src/objects/objects.cc
  Hunk #1 FAILED at 2491.
  Hunk #2 FAILED at 2517.
  2 out of 2 hunks FAILED -- saving rejects to file src/3rdparty/chromium/v8/src/objects/objects.cc.rej
```

**Comment by paperchalice** [comment]
From [official doc](https://doc.qt.io/qt-6/qtwebengine-platform-notes.html), it needs `html5lib` and `python3` so we need to create venv here and `qtwebengine` can be built on linux now.

**Comment by danielnachun** [comment]
I've tested this locally (without trying to build `qtwebengine`) and we will need to apply this upstream commit for it to build on Linux:
```
  # Apply upstream commit to fix build on older Linux systems
  patch do
    url "https://invent.kde.org/qt/qt/qtbase/-/commit/311d29d2261a7e0689340c4c1322138f8234da7b.diff"
    sha256 "a9cda53f71d307be9b35cbd60a56d369e5611dcf0bee3246c9f2f34b128928de"
    directory "qtbase"
  end
```
I'm going to try to build `qtwebengine` now and see what is needed for it to work on Linux (thanks for the pointers @paperchalice).

**Comment by danielnachun** [comment]
I am going to push a large commit that has the changes I needed to make for `qtwebengine` to build on Linux.  It will also have the fixes needed for the Qt SQL packages that were previously not working on Linux.  

Due to the very long build time and high disk usage I have not yet had a chance to test all of this on macOS, so please do not start a CI run until I've had chance to do this and make sure we don't waste a long run on something that fails several hours into the compilation.  The reason I'm pushing this now is to make it easier for me to check it out again later in a VM without having to copy over the changes manually.

A few things to note:
- Aside from the Python dependencies mentioned by @paperchalice, all of the commented out dependencies on Linux except `ffmpeg` are now needed (not counting `bluez` and `gypsy` which do not have formulae).  There were a few more dependencies needed as well but nothing big - just some X libraries.  I suspect a couple of these dependencies should also be used on macOS and will confirm if they get linked when I test it there.
- I opted to try to build `chromium` (used inside of `qtwebengine`) with "system" (Homebrew-installed) copies of third party dependencies instead of the vendored copies, for the dependencies which allow this.  `ffmpeg` still has to use the bundled copy for reasons that are explained in a comment in the PR.  `libjpeg` also needed to use the vendored copy, but I'm wondering if the issue there was that we need to use `jpeg-turbo` instead of `jpeg`, so I'm going to test that.
- An upstream commit was needed to fix using the "system" ICU, and I also had to make a small patch to address a rather arcane issue with some constants that were added the Linux 3.17 kernel but not copied into `glibc` until 2.27.  I'm not sure if upstream will accept this fix as Ubuntu 16.04 LTS isn't an officially supported platform, but we'll also be able to drop it once we move to Ubuntu 18.04 LTS anyway.
- I've tried unsuccessfully to fix `pyside` as well.  There was compilation issue that I have fixed, but now it does not properly install the python modules on LInux.  I don't want to hold this PR up to try to get that fixed.  However, whenever we do finally figure out how to build `pyside` on Linux, I think we should always make sure to sync its updates with `qt`.

**Comment by danielnachun** [comment]
I have pushed more commits to fix the build on macOS.  Now that we are not using Python 2 anymore, the Python resources for QtWebEngine are needed on macOS and Linux.  I also had to tweak the fix for finding the right clang path when building Chromium because it is still needed but is now in a different file.  I could not actually finish the build because it ran out of memory on macOS, but it came very close to the end so I'm reasonably confident it will finish in CI.

Looking at the build logs it seems that Chromium seems to always build the bundled copies of its dependencies on macOS.  This is not a good practice but I can't find a way to turn it off like on Linux.  For now I have moved the dependencies I had thought we would also need on macOS back to being Linux only. 

I would like to do one more Linux local test to see if `jpeg-turbo` will work instead of the bundled `libjpeg`, and once I've check that we can do a full CI run.

File: Formula/qt.rb [hunk 1]
```diff
@@ -111,17 +132,39 @@ class Qt < Formula
...
+  end
+
+  # Apply patch to fix chromium build with glibc < 2.27. See here for details:
+  # https://libc-alpha.sourceware.narkive.com/XOENQFwL/add-fcntl-sealing-interfaces-from-linux-3-17-to-bits-fcntl-linux-h
+
+  patch :DATA
+
   def install
+    # Install python dependencies for QtWebEngine
+    venv = virtualenv_create(libexec, "python3")
```
Comment (paperchalice) [inline Formula/qt.rb:156]: [Document](https://doc.qt.io/qt-6/qtwebengine-platform-notes.html#all-platforms) said `html5lib` is build time dependency.

Comment (danielnachun) [reply Formula/qt.rb:156]: Ah good catch there, this should go into a `venv` in `buildpath` instead.  This is pretty easy to test quickly as it checks if these dependencies are found very early in the build process.

Comment (danielnachun) [reply Formula/qt.rb]: I moved the `virtualenv` into `buildpath` and the dependencies were found successfully, so this is now fixed.

**Comment by danielnachun** [comment]
Building with `jpeg-turbo` does not fix the build failure encountered by Chromium when trying to use brewed `libjpeg` on Linux, so we have to use the vendored copy.  I believe this has been tested enough locally to do a full CI run that has a decent chance of success, so please feel free to start a CI run when a `CI-long-timeout` label is free.

**Comment by danielnachun** [comment]
All builds failed because of an issue with AVX2.  I hit this error locally on macOS and disabled AVX2 but was unsure if it was specific to my machine.  Clearly it's not so I'm going to disable AVX2 here on all platforms.

**Comment by paperchalice** [comment]
AVX issue may be related to [QTBUG-100817](https://bugreports.qt.io/browse/QTBUG-100817).

**Comment by danielnachun** [comment]
I think we are ready to go here.  There were only two test failures - `pympress` timed out on Intel Big Sur, and the `pdf-sandwich` failure is a known and unrelated problem with `imagemagick`: https://github.com/Homebrew/homebrew-core/pull/101936#issuecomment-1133511787.  

The only minor issue I've noticed here is that I should have used the `six` formula instead of a resource.  Since this is a build-only dependency it doesn't affect the bottle regardless of which one we use, so I don't think we should rerun CI for that, and can address this in the next revision bump or new version.

If there are no other thoughts I am going to publish the bottles in the next 24 hours.  I think 6.3.1 is coming out in just a few weeks so it will be good to get this out, and hopefully this next minor version will be a much simpler update!

**Review by danielnachun — APPROVED** [review]
I haven't seen any other feedback so I'm going to publish these bottles now.

**Pull request closed without merging**